### PR TITLE
Set az conversion

### PR DESCRIPTION
**Pull request by seongsujeong**
This PR is a replacement of #105 which was closed accidentally. 

Most of the comments / suggestions were applied in the previous PR. 

Below is my comment regarding the suggestion to separate `enu2rgaz` into `enu2rg` and `en2az`:
https://github.com/opera-adt/COMPASS/pull/105#discussion_r1157843023

### CODE REVIEW

File: src/compass/utils/geometry_utils.py [hunk 1]
```diff
@@ -265,3 +265,162 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+    3. Scale the normal vector in 2 with ENU displacement to
+       get the displacement in ECEF
+    4. Add the vectors calculated in 4 into 1.
+       This will be the ECEF coordinates after applying SET
+    5. Convert 4 into lat / lon / hgt.
+       This will be LLH coordinates after applying SET
+    6. Calculate the radar coordinate before SET applied using `geo2rdr`
+    7. Calculate the radar coordinate AFTER SET applied using `geo2rdr`
+    8. Calculate the difference between (7) and (6),
+       which will be the displacement in radargrid by SET
```
Comment (vbrancat) [inline src/compass/utils/geometry_utils.py:295]: I love it, thanks *-*

File: src/compass/utils/geometry_utils.py [hunk 2]
```diff
@@ -265,3 +265,162 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+        vec_e, vec_n, vec_u = get_enu_vector_ecef(lon_deg, lat_deg)
+
+        llh_ref = np.array([np.deg2rad(lon_deg),
+                            np.deg2rad(lat_deg),
+                            hgt])
+
+        xyz_before = ellipsoid.lon_lat_to_xyz(llh_ref)
+        xyz_after_set = (xyz_before + vec_e * e_arr[index_arr]
+                              + vec_n * n_arr[index_arr]
+                              + vec_u * u_arr[index_arr])
```
Comment (vbrancat) [inline src/compass/utils/geometry_utils.py:354]: It seems like an off formatting

File: src/compass/utils/geometry_utils.py [hunk 3]
```diff
@@ -265,3 +265,162 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+
+    Parameters
+    ----------
+    lon: np.ndarray
+        Longitude of the points to calculate ENU vectors
+    lat: np.ndarray
+        Latitude of the points to calculate ENU vectors
+    units: str
+        Units of the `lon` and `lat`.
+        Acceptable unit is `radian` or `degree` (Default)
```
Comment (vbrancat) [inline src/compass/utils/geometry_utils.py:395]:
```suggestion
        Acceptable unit is `radian` or `degree` (Default: degree)
```
Not sure, but I think units of angles need to be specified using the plural version?

**Review by vbrancat — COMMENTED** [review]
Thank you @seongsujeong, I did not have much time to test it (hopefully will do it soon). Just a couple of comments on formatting at the moment

**Comment by vbrancat** [comment]
@seongsujeong, while I was testing this PR, I noticed that the azimuth component of the Solid Earth Tides (SET) is not allocated inside the HDF5 CSLC product. However, it seems that the azimuth component of the correction is allocated inside the temporary product `corrections` in the scratch directory. Have you tested the PR? Can you reproduce the issue on your side?

**Comment by seongsujeong** [comment]
> @seongsujeong, while I was testing this PR, I noticed that the azimuth component of the Solid Earth Tides (SET) is not allocated inside the HDF5 CSLC product. However, it seems that the azimuth component of the correction is allocated inside the temporary product `corrections` in the scratch directory. Is this intentional? If not, I believe you should modify this function here
> 
> https://github.com/opera-adt/COMPASS/blob/8aeecda5f43b15869a0d044b78df9d7b9864d5b3/src/compass/utils/h5_helpers.py#L583
> and include the azimuth SET.

@vbrancat Thanks for catching this. I've made the changes on the function to add the azimuth component of SET LUT be written into the output HDF5.

**Review by vbrancat — APPROVED** [review]
LGTM, thanks :)

File: src/compass/utils/geometry_utils.py [hunk 4]
```diff
@@ -265,3 +265,163 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+             lon_arr, lat_arr, hgt_arr,
+             e_arr, n_arr, u_arr,
+             geo2rdr_params=None):
+    '''
+    Convert ENU displacement into range / azimuth displacement,
+    based on the idea mentioned in ETAD ATBD, available in the link below:
+    ("https://sentinels.copernicus.eu/documents/247904/4629150/"
+     "ETAD-DLR-DD-0008_Algorithm-Technical-Baseline-Document_2.3.pdf/"
+     "5cb45b43-76dc-8dec-04ef-ca1252ace434?t=1680181574715")
+
```
Comment (scottstanie) [inline src/compass/utils/geometry_utils.py:280]: Can we make this one line so we can click on it, maybe adding `# noqa` at the end if you are worried about it setting up linters?

File: src/compass/utils/lut.py [hunk 5]
```diff
@@ -266,13 +272,13 @@ def compute_geocoding_correction_luts(burst, dem_path, tec_path,
         wet_los_tropo = 2.0 * zen_wet / np.cos(np.deg2rad(inc_angle))
         dry_los_tropo = 2.0 * zen_dry / np.cos(np.deg2rad(inc_angle))
 
-    return geometrical_steering_doppler, bistatic_delay, az_fm_mismatch, [
-        rg_set, az_set], los_ionosphere, [wet_los_tropo, dry_los_tropo], los_static_tropo
+    return geometrical_steering_doppler, bistatic_delay, az_fm_mismatch,\
+           [rg_set, az_set], los_ionosphere,\
+            [wet_los_tropo, dry_los_tropo], los_static_tropo
```
Comment (scottstanie) [inline src/compass/utils/lut.py:277]:
```suggestion
    return (
        geometrical_steering_doppler, bistatic_delay, az_fm_mismatch,
        [rg_set, az_set], los_ionosphere,
        [wet_los_tropo, dry_los_tropo], los_static_tropo
    )
```
just a formatting suggestion so pylint doesn't say "line overindented for visual indent"

File: src/compass/utils/geometry_utils.py [hunk 6]
```diff
@@ -265,3 +265,163 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+                         'which needs to be either `degrees` or `radians`')
+
+    vec_u = np.array([np.cos(lon_rad) * np.cos(lat_rad),
+                      np.sin(lon_rad) * np.cos(lat_rad),
+                      np.sin(lat_rad)])
+
+    vec_n = np.array([-np.cos(lon_rad) * np.sin(lat_rad),
+                      -np.sin(lon_rad) * np.sin(lat_rad),
+                      np.cos(lat_rad)])
+
```
Comment (scottstanie) [inline src/compass/utils/geometry_utils.py:424]: If you've got some easy reference for how you wrote these, that would be helpful here

File: src/compass/utils/geometry_utils.py [hunk 7]
```diff
@@ -265,3 +265,163 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
         ]
 
     return unit_vec
+
+
```
Comment (scottstanie) [inline src/compass/utils/geometry_utils.py:269]: A comment more on the `geometry_utils.py` module which might be for @vbrancat 
```
$ grep "def " src/compass/utils/geometry_utils.py | cut -d ' ' -f2 | cut -d'(' -f1  # list of functions defined here
los2orbit_azimuth_angle
azimuth2heading_angle
heading2azimuth_angle
enu2los
en2az
calc_azimuth_from_east_north_obs
get_unit_vector4component_of_interest
enu2rgaz
get_enu_vector_ecef
```

of these 9 fucntions, we're not using 5 of them, at least in the current state of this PR:




- azimuth2heading_angle
- enu2los
- en2az
- calc_azimuth_from_east_north_obs
- get_unit_vector4component_of_interest

Should we get rid of them until we're actually needing the function?

(how i checked they're not used:)
<details>

```bash
COMPASS$ grep "def " src/compass/utils/geometry_utils.py | cut -d ' ' -f2 | cut -d'(' -f1 | xargs -n1 git grep
src/compass/utils/geometry_utils.py:def los2orbit_azimuth_angle(los_az_angle, look_direction='right'):
src/compass/utils/geometry_utils.py:        orb_az_angle = los2orbit_azimuth_angle(los_az_angle)
src/compass/utils/geometry_utils.py:def azimuth2heading_angle(az_angle, look_direction='right'):
src/compass/utils/geometry_utils.py:def heading2azimuth_angle(head_angle, look_direction='right'):
src/compass/utils/geometry_utils.py:            az_angle = heading2azimuth_angle(head_angle)
src/compass/utils/geometry_utils.py:def enu2los(v_e, v_n, v_u, inc_angle, head_angle=None, az_angle=None):
src/compass/utils/geometry_utils.py:def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu2los', horz_az_angle=None):
src/compass/utils/geometry_utils.py:        enu2los, en2los, hz2los, u2los, up2los, orb(it)_az, vert, horz
src/compass/utils/geometry_utils.py:        'enu2los', 'en2los', 'hz2los', 'horz2los', 'u2los', 'vert2los',   # radar LOS / cross-track
src/compass/utils/geometry_utils.py:    if comp in ['enu2los']:
src/compass/utils/geometry_utils.py:def en2az(v_e, v_n, orb_az_angle):
src/compass/utils/geometry_utils.py:        'en2az', 'hz2az', 'orb_az', 'orbit_az',                           # radar azimuth / along-track
src/compass/utils/geometry_utils.py:    elif comp in ['en2az', 'hz2az', 'orb_az', 'orbit_az']:
src/compass/utils/geometry_utils.py:def calc_azimuth_from_east_north_obs(east, north):
src/compass/utils/geometry_utils.py:def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu2los', horz_az_angle=None):
src/compass/utils/geometry_utils.py:def enu2rgaz(radargrid_ref, orbit, ellipsoid,
src/compass/utils/lut.py:from compass.utils.geometry_utils import enu2rgaz
src/compass/utils/lut.py:    set_rg, set_az = enu2rgaz(burst.as_isce3_radargrid(), burst.orbit, ellipsoid,
src/compass/utils/geometry_utils.py:        vec_e, vec_n, vec_u = get_enu_vector_ecef(lon_deg, lat_deg)
src/compass/utils/geometry_utils.py:def get_enu_vector_ecef(lon, lat, units='degrees'):
```
</details>

**Review by scottstanie — COMMENTED** [review]
Going to go try testing this now- do you have any checks that are useful to look for? What did you do so far to test it?

File: src/compass/utils/geometry_utils.py [hunk 8]
```diff
@@ -265,3 +265,163 @@ def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu
...
+    Algorithm description
+    ---------------------
+    For all lon / lat / height of the array;
+    1. Calculate the ECEF coordinates before applying SET
+    2. Calculate the normal vector of east / north / up directions of the point
+    3. Scale the normal vector in 2 with ENU displacement to
+       get the displacement in ECEF
+    4. Add the vectors calculated in 4 into 1.
+       This will be the ECEF coordinates after applying SET
```
Comment (scottstanie) [inline src/compass/utils/geometry_utils.py:289]: shoot my comment here didn't get attached to the other
2. "calculate the normal vector of ..." Should this be `the unit vector`? Normal means perpendicular to something
3. Same here for `normal`. What are we scaling the vector by? should be "transform" or "reproject"?
4. `Add the vectors calculated in 4 into 1.` this is step 4, is it add the vectors in (3)?

**Comment by vbrancat** [comment]
@scottstanie it would be great if you can have a look at this PR. Thanks

**Comment by scottstanie** [comment]
1. Looks like changes to isce v0.12 means it's erroring because of `azimuth_carrier` being the wrong name for an argument:
```
  File "/u/aurora-r0/staniewi/repos/sweets/src/sweets/_geocode_slcs.py", line 59, in run_geocode
    s1_geocode_slc.run(cfg)
  File "/u/aurora-r0/staniewi/repos/COMPASS/src/compass/s1_geocode_slc.py", line 158, in run
    isce3.geocode.geocode_slc(geo_burst_raster, rdr_burst_raster,
TypeError: geocode_slc() got an unexpected keyword argument 'azimuth_carrier'
```
2. any chance of adding a unit test for some of the new/changed functionality?

File: src/compass/utils/geometry_utils.py (hunk 7 above)
Comment (seongsujeong) [reply src/compass/utils/geometry_utils.py:269]: I've intentionally left the functions used for previous implementation about the conversion, which is based on radar geometry. Maybe it's a good idea to take off the unused functions before getting this PR merged. What do you think @vbrancat?

File: src/compass/utils/geometry_utils.py (hunk 6 above)
Comment (seongsujeong) [reply src/compass/utils/geometry_utils.py:425]: Good idea. In fact I did my own calculations, but I've added couple of links, which is essentially similar to what I did.

File: src/compass/utils/geometry_utils.py (hunk 8 above)
Comment (seongsujeong) [reply src/compass/utils/geometry_utils.py]: Good catch. I've updated the description in the docstring.

File: src/compass/utils/geometry_utils.py (hunk 7 above)
Comment (vbrancat) [reply src/compass/utils/geometry_utils.py:269]: My preference is to not remove them. We want to keep some functionalities down the line they might be useful. Anyhow, which of these functions are not used?

Comment (seongsujeong) [reply src/compass/utils/geometry_utils.py:269]: > My preference is to not remove them. We want to keep some functionalities down the line they might be useful. 
Sounds good to me. I will keep those functions then.

> Anyhow, which of these functions are not used?
I meant the 5 functions that @scottstanie has mentioned above, and also was thinking about checking the codes to see if there are more.

**Comment by seongsujeong** [comment]
@scottstanie I'm not sure if the errors you reported in the comment [above](https://github.com/opera-adt/COMPASS/pull/132#issuecomment-1542680642) is related to this PR. I know it can be a bit tricky to test this PR with the most up-to-date ISCE3, but if the change look good to you overall?

**Review by scottstanie — APPROVED** [review]

**Pull request merged**